test/mpi: misc fixes and enhancements in test/mpi

ggouaillardet opened the pull request:
This PR fixes misc issues and add a few options so
make dist
can be ran from the test/mpi directory, and the tarball can be used standalone, for example with OpenMPI

raffenet commented:
test:jenkins/ch3/tcp

raffenet commented:
@ggouaillardet In 65bfe38, you need to remove `f77tof90` from the substitutions in `maint/configure.ac` else `autogen.sh` won't run all the required scripts. Once you update I will restart our integration testing.

ggouaillardet commented:
test:jenkins/ch3/tcp

ggouaillardet commented:
@raffenet i made the requested changes, plus some related cleanup

raffenet commented:
Thanks! I'll test it out and review sometime this week.

test:jenkins/ch3/tcp

raffenet commented on test/mpi/Makefile.am:
```diff
@@ -19,9 +19,9 @@ all_lang_subdirs = f77 cxx f90 f08
 # assignments that are not driven by a configure @-substitution variable.
 DIST_SUBDIRS = $(static_subdirs) io $(all_lang_subdirs) threads spawn .
-SUBDIRS      = $(static_subdirs) $(iodir) $(otherlangs) $(threadsdir) $(spawndir) $(ckpointdir) $(ftdir) .
+SUBDIRS      = $(static_subdirs) $(iodir) $(otherlangs) $(threadsdir) $(spawndir) $(ckpointdir) ft .
```
I think another patch may have been accidentally merged into this one...

ggouaillardet replied:
that was intended.

if ```SUBDIRS``` does not contain ```ft```, and if the test suite is configure'd with ```--disable-ft-tests```, then ```make dist``` will fail because of the missing ```ft``` directory.
so i think the right approach is to always include the ```ft``` directory, and __not__ build the ft related tests if they are disabled.
see the full commit at ggouaillardet/mpich@441853771c322278b9c819c001eeb2998cc53aa7

and i should have applied the same logic to all the directories.
if that makes sense, i will do so and update the PR

ggouaillardet commented on test/mpi/configure.ac:
```diff
@@ -380,12 +380,7 @@ if test "$enable_strictmpi" = "yes" -o "$enable_perftest" = "no" ; then
-    ftdir="ft"
-fi
-AC_SUBST(ftdir)
+AM_CONDITIONAL(BUILD_FT, [test "$enable_ft_tests" = "xyes"])
```
that was mistakenly oversimplified, will fix too

ggouaillardet commented:
@raffenet i just pushed some changes related to ```ft-tests```
that will hopefully fix the errors reported by Jenkins

could you please restart the integration testing ?

pavanbalaji commented:
Looks like this PR got dropped somehow.  Restarting testing.

test:jenkins/ch3/tcp

pavanbalaji commented:
This PR seems to be broken.  @ggouaillardet would you be able to fix this?

ggouaillardet commented:
@pavanbalaji i updated and rebased the PR, it will hopefully fix the `@ftdir@` related failures.

could you please have jenkins retest this ?

pavanbalaji commented:
test:jenkins/ch3/tcp

ggouaillardet commented:
Only one test failed (`RMA` in `freebsd32`) and the same test fails in most PRs (but in `freebsd64`).
I am not sure about what this is about, but at first glance, this seems unrelated with this PR.

Meanwhile, a conflict appeared, so I updated the PR.

@pavanbalaji could you please have jenkins retest this ?

pavanbalaji commented:
If it's the connection error on FreeBSD, that's a slurm issue that we are looking into.  You can ignore those for now.

test:jenkins/ch3/tcp

ggouaillardet commented:
@pavanbalaji I rebased this PR and fixed the conflicts

raffenet commented:
@ggouaillardet thanks! One of us will take a look soon.

raffenet commented:
Still working on reviewing a few patches. I've made a few changes, so triggering the tests to make sure I didn't break anything.
test:jenkins/ch3/tcp

raffenet commented on test/mpi/configure.ac:
```diff
@@ -186,6 +186,11 @@ AC_ARG_ENABLE(profile-tests,
 		[Include f90 for PMPI (default)])],,
 	[enable_profile_tests=yes])
 
+AC_ARG_ENABLE(f90-sizeof-tests,
```
@ggouaillardet what's the reason these tests might not build? I'm mostly just curious, but it would be good to document in the commit message or elsewhere.

ggouaillardet replied:
@raffenet I rebased the PR vs the latest master in order to fix a (trivial) conflict.

I updated the commit message to explain the rationale for this option

> MPI_Sizeof() requires some level of support by the Fortran compiler.
> For example, Open MPI does not provide this subroutine when built
> on a RHEL 7 box with the stock gfortran (4.8.5) compiler.

It would probably best to remove this option and have `configure` automagically figure out if `MPI_Sizeof()` (and possibly the `ft` and `profile` tests too) can be used.

Do you wish me moving forward and implement this ?

ggouaillardet commented:
@raffenet I revamped the `configure` logic and now autodetects support for `ft`, `profile` and `MPI_Sizeof()` tests. I also automatically detects implementation that provides Fortran 2008 bindings.

I noted the detection of `MPI_2DOUBLE_COMPLEX` and `MPI_2COMPLEX` from within MPICH always failed, so I forced that when `FROM_MPICH` is `yes`

raffenet commented:
Actually it seems the 2COMPLEX type in MPICH is not defined without additional CFLAGS. I plan to drop that patch once these tests complete, as well.

ggouaillardet commented:
@raffenet I fixed two issues
 - `configure` used to fail tests for `MPI_2COMPLEX` and `MPI_2DOUBLE_COMPLEX` but for a bad reason (missing `mpi.h`), but they should generally fail, so do not test these datatypes when configuring from MPICH. I will improve this later.
- I thought I was the one who introduced the `--enable-ft-tests` option, but I was wrong.
  There were some things that should have not be dropped from my commit (so `make dist` success even if ft tests cannot be built), and I added the fixes into a previous `ft` related commit

Could you please have jenkins retest this PR ?

raffenet commented on test/mpi/configure.ac:
```diff
@@ -1438,6 +1438,10 @@ fi
 if test "$pac_cv_have_mpi_integer16" = yes ; then 
     AC_DEFINE(HAVE_MPI_INTEGER16,1,[Define if MPI_INTEGER16 is available])
 fi
+if test "$FROM_MPICH" = yes ; then
```
Why bother with this instead of relying on the test below? Can this patch be dropped?

ggouaillardet replied:
Unfortunately, you cannot ... `#include <mpi.h>` always fail from MPICH and regardless the appropriate flags were passed or not. I fixed my last commit (and rebased everything), and hopefully, we should be all set (and the last commit will be kept)

ggouaillardet replied:
If you run `configure` from the MPICH top directory, here is what you currently find in `test/mpi/config.log`

```
configure:12304: checking whether MPI_2DOUBLE_COMPLEX is available
configure:12325: gcc -std=gnu99 -c    conftest.c >&5
conftest.c:52:25: fatal error: mpi.h: No such file or directory
         #include "mpi.h"
                         ^
compilation terminated.
configure:12325: $? = 1
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "mpich-testsuite"
| #define PACKAGE_TARNAME "mpich-testsuite"
| #define PACKAGE_VERSION "3.3b3"
| #define PACKAGE_STRING "mpich-testsuite 3.3b3"
| #define PACKAGE_BUGREPORT "discuss@mpich.org"
| #define PACKAGE_URL "http://www.mpich.org/"
| #define PACKAGE "mpich-testsuite"
| #define VERSION "3.3b3"
| #define HAVE_MPI_SPAWN 1
| #define HAVE_MPI_WIN_CREATE 1
| #define HAVE_LONG_DOUBLE 1
| #define HAVE_LONG_LONG 1
| #define restrict __restrict
| #define STDC_HEADERS 1
| #define HAVE_SYS_TYPES_H 1
| #define HAVE_SYS_STAT_H 1
| #define HAVE_STDLIB_H 1
| #define HAVE_STRING_H 1
| #define HAVE_MEMORY_H 1
| #define HAVE_STRINGS_H 1
| #define HAVE_INTTYPES_H 1
| #define HAVE_STDINT_H 1
| #define HAVE_UNISTD_H 1
| #define HAVE_STDARG_H 1
| #define HAVE_SYS_TIME_H 1
| #define HAVE_SYS_RESOURCE_H 1
| #define HAVE__BOOL 1
| #define HAVE_FLOAT__COMPLEX 1
| #define HAVE_DOUBLE__COMPLEX 1
| #define HAVE_LONG_DOUBLE__COMPLEX 1
| #define SIZEOF_VOID_P 8
| #define USE_LONG_DOUBLE_COMPLEX 1
| #define HAVE_PTHREAD_H 1
| #define HAVE_PTHREAD_CREATE 1
| #define HAVE_PTHREAD_YIELD 1
| #define HAVE_PTHREAD_BARRIER_INIT 1
| #define HAVE_PTHREAD_BARRIER_WAIT 1
| #define HAVE_H_ADDR_LIST 1
| #define HAVE_GETRUSAGE 1
| #define HAVE_FORTRAN_BINDING 1
| #define STDCALL
| #define F77_NAME_LOWER_USCORE 1
| #define HAVE_IOSTREAM 1
| #define HAVE_NAMESPACES /**/
| #define HAVE_NAMESPACE_STD /**/
| #define HAVE_MPI_IO 1
| #define MPIO_USES_MPI_REQUEST /**/
| #define HAVE_MPI_INTEGER16 1
| /* end confdefs.h.  */
|
|         #include "mpi.h"
| int
| main ()
| {
|
|             MPI_Datatype t = MPI_2DOUBLE_COMPLEX;
|
|   ;
|   return 0;
| }
|
configure:12337: result: no
```

the test fail (this is fine since `MPI_2DOUBLE_COMPLEX` is not available out of the box) but for a bad reason (`fatal error: mpi.h: No such file or directory`), and it will always fail when invoked from MPICH. That is why, from MPICH, we cannot use `mpi.h` but we should use the macro directly.

makes sense ?

raffenet replied:
Yes, now I understand the problem. I still don't like the solution, since the underlying problem is not solved. I still prefer to drop this patch and work on a solution to properly accessing `mpi.h`.

pavanbalaji commented:
Subsumed by https://github.com/pmodels/mpich/pull/4523
